### ISSUE TITLE
question: Localstack slowed down a lot after Edge Service introduced, is that known/acknowledged?

### ISSUE THREAD

**Issue by frncmx-form3** [localstack/localstack#4341]
### Is there an existing issue for this?

- [X] I have searched the existing issues and read the documentation

### Question

**Our setup**

We have a huge integration test suite which runs for 39 minutes on average on Travis. (That's too much, but that's a different topic.) We run Localstack in a container and we provision it with Terraform.

The tests are SQS heavy, so the issue might be related to: https://github.com/localstack/localstack/issues/3998

The test suite used a very old Localstack version (`v0.10.6`), so I recently started to upgrade.

**The issue**

So `v0.10.6` still used ElasticMQ as SQS backend and no Edge service. As our baseline, the test suite run for 39 minutes on that. When I bumped the version to `v0.12.15` (latest atm), that build slowed down significantly (10+ minutes). So, I started to investigate.

I observed the following:
| Localstack version | SQS backend | Edge Service | Runtime | Delta |
| ------------------------ | ------------------- | ------------------ | -------------------------- | ------- |
| `v0.11.2` | ElasticMQ (default) | no | 39 m | 0 | 
| `v0.11.2` | ElasticMQ (default) | yes | 51 m | +12 | 
| `v0.11.3` | Moto (default) | no | 50 m | +11 m | 
| `v0.11.3` | ElasticMQ | no | 44 m | +5 m | 
| `v0.12.15` | Moto (default) | yes | 60 m | +21 m |
| `v0.12.15` | ElasticMQ | yes | 50 m | +11 m |

**Questions**

1) Is this something you are aware of?
2) Are you actively looking into improving performance?
3) Is there anything I could try to tweak to improve performance?

(I did not dig into this yet any deeper than the above presented table. At the moment, we decide to stick with Localstack version `v0.11.2`.)

**Comment by seibelj** [comment]
The latest version of localstack (0.12.15) is extremely slow. Lambda requests take 8 to 10 seconds rather than 1 to 2. We have reverted back to `image: localstack/localstack-full:0.12.7` for stabilty.

**Comment by NeilJ247** [comment]
It's massively slow.  I have seen tests for SQS processing jump to 7+ minutes to run and even then I don't think it has completed the queue processing due to failing assertions that passed in previous builds.

**Comment by mariusingjer** [comment]
I stumbled upon this thread will debugging some abysmal performance when publishing to sqs backed by moto 

(https://gist.github.com/mariusingjer/1011853931d7d4f68cf0891930094df4).

I changed the sqs provider from moto to elasticmq in my localstack docker:

```yaml
  localstack:
    image: localstack/localstack:0.13.1
    ports:
      - "4566-4584:4566-4584"
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}"
    environment:
      - SERVICES=s3,sqs,sns,events
      - PORT_WEB_UI=8080
      - DOCKER_HOST=unix:///var/run/docker.sock
      - DEFAULT_REGION=us-east-1
      - DEBUG=0,
      - SQS_PROVIDER=elasticmq
    volumes:
      - "localstack-vol:/tmp/localstack"
```
And I could run the same tests again, with linear (and fast) publishing times

**Comment by whummer** [comment]
Thanks for the update @mariusingjer . In fact, we have recently shipped our own SQS implementation (in #5148 ) due to the large number of performance issues that have been reported in the past. 

You can enable this new provider by setting `PROVIDER_OVERRIDE_SQS=asf` in the LocalStack environment configuration ("asf" stands for "AWS Service Framework", a framework we're developing to provide custom service implementations). Over time, this custom implementation ("asf" provider) will likely become the default in LocalStack.

So, to summarize - if you come across this issue (performance issues with SQS), then we recommend either:

(1) Configuring the ElasticMQ provider:
```
SQS_PROVIDER=elasticmq
```
or, (2) Trying out our new custom implementation by configuring:
```
PROVIDER_OVERRIDE_SQS=asf
```

Hope that helps - please keep us posted how it goes. We'll continue to invest heavily in further improving the performance of the edge proxy, as well as the individual services. More updates to follow soon! Thanks for your help.
